### PR TITLE
rate limit filter

### LINKED ISSUE #10: Add rate limit filter
**What would you like to be added**:
A filter can control the request rate, sentinel-go may be a good idea. 

**Why is this needed**:
Rate limit avoid heavy traffic make back-end service not available.

### PR DESCRIPTION
**Pull request by mark4z**
**What this PR does**:

**Which issue(s) this PR fixes**:

Fixes #10

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/filter/ratelimit/init.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+)
+
+import (
+	"github.com/apache/dubbo-go-pixiu/pkg/filter/ratelimit/config"
+	"github.com/apache/dubbo-go-pixiu/pkg/filter/ratelimit/matcher"
+	"github.com/apache/dubbo-go-pixiu/pkg/logger"
+)
+
+func GetConfig() *config.Config {
+	c, _ := GetMockedRateLimitConfig()
```
Comment (williamfeng323) [inline pkg/filter/ratelimit/init.go:35]: i think you might call the wrong function.

Comment (mark4z) [reply pkg/filter/ratelimit/init.go:35]: yep, that func does not yet exist. i'll do it later

**Review by AlexStocks — APPROVED** [review]

**Comment by AlexStocks** [comment]
pls fix the go.mod confliction.

**Review by williamfeng323 — APPROVED** [review]

**Review by AlexStocks — APPROVED** [review]

File: pkg/filter/plugins/plugins.go [hunk 2]
```diff
@@ -165,7 +189,7 @@ func getAPIFilterFuncsWithPluginsGroup(plu *config.PluginsInUse) []context.Filte
 
 	// found with group name
 	for _, groupName := range plu.GroupNames {
-		pwfMap, found := groupWithPluginsMap[groupName]
+		pwfMap, found := groupCache[groupName]
 		if found {
```
Comment (zhangshen023) [inline pkg/filter/plugins/plugins.go:193]: `	if !found {
		continue
	}
	for _, pwf := range pwfMap {
		tmpMap[pwf.Name] = pwf.fn
	}
`

File: pkg/filter/ratelimit/logger_warpper.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
+/*
```
Comment (zhangshen023) [inline pkg/filter/ratelimit/logger_warpper.go:1]: what does this code make sense?

File: pkg/filter/ratelimit/matcher/exact.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+)
+
+import (
+	"sync"
+)
+
+type Exact struct {
+	apiNames map[string]string
+
+	mu sync.RWMutex
```
Comment (zhangshen023) [inline pkg/filter/ratelimit/matcher/exact.go:31]: FYI,I think pointer is better

**Review by zhangshen023 — APPROVED** [review]

File: pkg/filter/ratelimit/matcher/exact.go (hunk 4 above)
Comment (mark4z) [reply pkg/filter/ratelimit/matcher/exact.go:31]: mu *sync.RWMutex, this way?

**Pull request merged**